tests/test_cli.py

```diff
@@ -383,4 +381,155 @@ def test_start_server(managed_tmpdir):
         res = runner.invoke(cli.server, ['--ip', 'localhost', '--port', '50111', '--timeout', '1'])
         assert time.time() - startTime <= 1.8  # buffer to give it time to stop
         assert res.exit_code == 0
-        assert 'Hangar Server Started' in res.stdout
\ No newline at end of file
+        assert 'Hangar Server Started' in res.stdout
+
+
+# =========================== External Plugin =================================
+
+def monkeypatch_scan(provides, accepts, attribute, func):
+    def wrapper(self):
+        from hangar.external import BasePlugin
+
+        plugin = BasePlugin(provides, accepts)
+        plugin.__dict__[attribute] = func
+
+        self._plugin_store['myplugin'] = plugin
+    return wrapper
+
+
+def test_import(monkeypatch, written_repo):
+    from hangar.external import PluginManager
+    repo = written_repo
+    runner = CliRunner()
+    shape = (5, 7)
+    fpath = 'data.ext'
+    aset_name = 'writtenaset'
+
+    def load(fpath, *args, **kwargs):
+        data = np.random.random((5, 7)).astype(np.float64)
+        return data, fpath
+
+    with monkeypatch.context() as m:
+        m.setattr(PluginManager, "_scan_plugins", monkeypatch_scan(['load'], ['ext'], 'load', load))
+
+        # invalid branch
+        res = runner.invoke(cli.import_data, [aset_name, fpath, '--branch', 'invalid'], obj=repo)
+        assert 'invalid does not exist' in res.stdout
+
+        # invalid plugin
+        res = runner.invoke(cli.import_data, [aset_name, fpath, '--plugin', 'invalid'], obj=repo)
+        assert str(res.exception) == 'Plugin invalid not found'
+
+        # adding data
+        res = runner.invoke(cli.import_data, [aset_name, fpath], obj=repo)
+        assert res.exit_code == 0
+        co = repo.checkout(write=True)
+        co.commit('added data')
+        d1 = co.arraysets[aset_name][fpath]
+        co.close()
+
+        # without overwrite
+        res = runner.invoke(cli.import_data, [aset_name, fpath], obj=repo)
+        assert res.exit_code == 0
+        co = repo.checkout()
+        d2 = co.arraysets[aset_name][fpath]
+        assert np.allclose(d1, d2)
+
+        # with overwrite
+        res = runner.invoke(cli.import_data, [aset_name, fpath, '--overwrite'], obj=repo)
+        assert res.exit_code == 0
+        co = repo.checkout(write=True)
+        co.commit('added data')
+        d3 = co.arraysets[aset_name][fpath]
+        co.close()
+        assert not np.allclose(d1, d3)
```

Comment:
Move all individual `"checks"` into their own tests. It's not apparent that these are individual test cases, and that the first two (should) fail, the third is the "garden path", and the last two test optional flags/behavior. 

setting up the repo as a fixture and using named tests would be much clearer here.

Reply:
I was trying to save few seconds (setup time) here but it makes sense to keep it your way for readability

Reply:
Ok. thanks!